### PR TITLE
Feature/restart pods on sidecar changes

### PR DESCRIPTION
**Pull request by kvahed**
added side car changes to pod rotation triggering

### CODE REVIEW

File: tests/sidecar_test.go [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+		depl.Coordinators.Sidecars = spec.GroupSideCars(grp)
+	})
+	if err != nil {
+		t.Fatalf("Failed to delete all containers from both coordinators and db servers", grp)
+	} else {
+		t.Log("Updated deployment")
+	}
+
+	removeDeployment(c, depl.GetName(), ns)
+
```
Comment (joerg84) [inline tests/sidecar_test.go]: Just curious: what are our whitespace style guide rules

Comment (maierlars) [reply tests/sidecar_test.go]: we use `go fmt`.

File: pkg/apis/deployment/v1alpha/deployment_status.go [hunk 2]
```diff
@@ -65,6 +68,9 @@ type DeploymentStatus struct {
 
 	// ForceStatusReload if set to true forces a reload of the status from the custom resource.
 	ForceStatusReload *bool `json:"force-status-reload,omitempty"`
+
+	// SideCarSpecs contains list of specifications specified for side cars
+	SideCarSpecs map[string]v1.Container
```
Comment (maierlars) [inline pkg/apis/deployment/v1alpha/deployment_status.go]: Why is this in the `DeploymentStatus`?

File: pkg/apis/deployment/v1alpha/member_status.go [hunk 3]
```diff
@@ -64,7 +67,8 @@ func (s MemberStatus) Equal(other MemberStatus) bool {
 		s.PodName == other.PodName &&
 		s.Conditions.Equal(other.Conditions) &&
 		s.IsInitialized == other.IsInitialized &&
-		s.CleanoutJobID == other.CleanoutJobID
+		s.CleanoutJobID == other.CleanoutJobID &&
+		reflect.DeepEqual(s.SideCarSpecs, other.SideCarSpecs)
```
Comment (maierlars) [inline pkg/apis/deployment/v1alpha/member_status.go]: @joerg84 mentioned a library for deep compare for kubernetes native types

File: pkg/util/k8sutil/images.go [hunk 4]
```diff
@@ -67,3 +67,17 @@ func GetArangoDBContainerFromPod(pod *corev1.Pod) corev1.Container {
 
 	return arangoc
 }
+
+// GetNonArangoNonExporterFromPod returns all non ArangoDB/Exporter containers from a pod
+func GetOtherContainersFromPod(pod *corev1.Pod) corev1.Container {
```
Comment (maierlars) [inline pkg/util/k8sutil/images.go]: Comment and Function name does not match.

File: pkg/util/k8sutil/images.go [hunk 5]
```diff
@@ -67,3 +67,17 @@ func GetArangoDBContainerFromPod(pod *corev1.Pod) corev1.Container {
...
+	arangoc := pod.Spec.Containers[0]
+	if len(pod.Status.ContainerStatuses) > 1 {
+		for _, container := range pod.Spec.Containers {
+			if strings.Contains(container.Name, "server") {
+				arangoc = container
+			}
+		}
+	}
```
Comment (maierlars) [inline pkg/util/k8sutil/images.go]: This function is just copy paste and does not do what the comment claims.

File: pkg/deployment/reconcile/plan_builder.go [hunk 6]
```diff
@@ -394,9 +395,109 @@ func podNeedsRotation(log zerolog.Logger, p v1.Pod, apiObject metav1.Object, spe
...
+		}
+		if !found {
+			return true, "Sidecar " + name + " no longer in specification"
+		}
+	}
+
 	return false, ""
 }
 
+func sideCarRequireRotation(wanted, given *v1.Container) bool {
```
Comment (maierlars) [inline pkg/deployment/reconcile/plan_builder.go]: missing comment

File: pkg/apis/deployment/v1alpha/deployment_status.go (hunk 2 above)
Comment (kvahed) [reply pkg/apis/deployment/v1alpha/deployment_status.go]: left over. good catch

File: tests/test_util.go [hunk 7]
```diff
@@ -558,6 +560,54 @@ func createEqualVersionsPredicate(version driver.Version) func(driver.VersionInf
...
+		// How many pods not matching
+		noGood = 0
+
+		// Check member after another
+		apiObject.ForeachServerGroup(func(group api.ServerGroup, spec api.ServerGroupSpec, status *api.MemberStatusList) error {
+
+			t.Logf("%s group has %d sidecars", group.AsRole(), spec.GetSidecars())
+
+			for _, m := range *status {
+				for _, scar := range spec.GetSidecars() {
```
Comment (joerg84) [inline tests/test_util.go]: What happens if we remove sidecars? I.e., don't we have to check the reverse `are all m.sidecarspecs also in the spec`? E.g., len(SideCarSpecs) == len(spec.GetSidecars()) or something similar?

File: tests/test_util.go [hunk 8]
```diff
@@ -558,6 +560,54 @@ func createEqualVersionsPredicate(version driver.Version) func(driver.VersionInf
...
 }
 
+// clusterSidecarsEqualSpec returns nil if sidecars from spec and cluster match
+func waitUntilClusterSidecarsEqualSpec(t *testing.T, spec api.DeploymentMode, depl api.ArangoDeployment) error {
+
+	c := cl.MustNewInCluster()
+	ns := getNamespace(t)
+
+	var noGood int
+	for start := time.Now(); time.Since(start) < 300*time.Second; {
```
Comment (joerg84) [inline tests/test_util.go]: This feels very weird... 
Shouldn't we  use something similar to: https://github.com/arangodb/kube-arangodb/blob/cf338bd90649c65662ce6eedbc98fd84c50b8def/pkg/util/retry/retry.go#L124 (or L137 in the same File)?

File: tests/test_util.go [hunk 9]
```diff
@@ -558,6 +560,54 @@ func createEqualVersionsPredicate(version driver.Version) func(driver.VersionInf
...
+		// Check member after another
+		apiObject.ForeachServerGroup(func(group api.ServerGroup, spec api.ServerGroupSpec, status *api.MemberStatusList) error {
+
+			t.Logf("%s group has %d sidecars", group.AsRole(), spec.GetSidecars())
+
+			for _, m := range *status {
+				for _, scar := range spec.GetSidecars() {
+					mcar, found := m.SideCarSpecs[scar.Name]
+					if found {
+						if !reflect.DeepEqual(mcar, scar) {
```
Comment (joerg84) [inline tests/test_util.go]: sideCarRequireRotation above seems to require a more complex comparison. Is reflect.DeepEqual sufficient here?

File: tests/sidecar_test.go [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+	spec.AddSideCar(grp, container)
+
+	deployment, err = updateDeployment(c, depl.GetName(), ns,
+		func(depl *api.DeploymentSpec) {
+			depl.Coordinators.Sidecars = spec.GroupSideCars(grp)
+		})
+	if err != nil {
+		t.Fatalf("Failed to add %s to group %s", container.Name, grp)
+	} else {
+		t.Log("XXXXX")
```
Comment (joerg84) [inline tests/sidecar_test.go]: Please don't forget to update

File: tests/sidecar_test.go [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+	return fmt.Sprintf("%s:%s", imageName, imageTag)
+}
+func (s *sideCarTest) ClearGroupSideCars(group string) {
+	s.sideCars[group] = nil
+}
+
+// TestSideCars tests side car functionality
+func TestSideCars(t *testing.T) {
+	runSideCarTest(t, &sideCarTest{
+		version: "3.4.7",
```
Comment (maierlars) [inline tests/sidecar_test.go]: Does the test always use `arangodb/arangodb:3.4.7`?

File: tests/test_util.go (hunk 9 above)
Comment (neunhoef) [reply tests/test_util.go]: For the test this is sufficient here, I think.

File: tests/test_util.go (hunk 7 above)
Comment (neunhoef) [reply tests/test_util.go]: I think this might be a valid comment.

File: tests/test_util.go (hunk 8 above)
Comment (neunhoef) [reply tests/test_util.go]: Possible, but is it worth doing?

File: tests/test_util.go (hunk 7 above)
Comment (kvahed) [reply tests/test_util.go]: fair point

File: tests/sidecar_test.go (hunk 11 above)
Comment (kvahed) [reply tests/sidecar_test.go]: until changed yes

File: tests/test_util.go (hunk 8 above)
Comment (kvahed) [reply tests/test_util.go]: is offered under good practices

File: tests/test_util.go (hunk 9 above)
Comment (kvahed) [reply tests/test_util.go]: yes is sufficient for the tests

**Pull request merged**